engines/rados.c

```diff
@@ -0,0 +1,501 @@
+	if (r < 0) {
+		log_err("rados_conf_read_file failed.\n");
+		goto failed_early;
+	}
+
+	r = rados_connect(rados->cluster);
+	if (r < 0) {
+		log_err("rados_connect failed.\n");
+		goto failed_early;
+	}
+
+	r = rados_ioctx_create(rados->cluster, o->pool_name, &rados->io_ctx);
+	if (r < 0) {
+		log_err("rados_ioctx_create failed.\n");
+		goto failed_shutdown;
+	}
+
+	for (i = 0; i < rados->object_count; i++) {
+		f = td->files[i];
+		f->real_file_size = file_size;
+		f->engine_pos = i;
+
+		oname_len = strlen(f->file_name) + 32;
+		rados->objects[i] = malloc(oname_len);
+		// vary objects for different jobs
+		snprintf(rados->objects[i], oname_len - 1,
+			"fio_rados_bench.%s.%x",
+			f->file_name, td->thread_number);
+		r = rados_write(rados->io_ctx, rados->objects[i], "", 0, 0);
+		if (r < 0) {
+			free(rados->objects[i]);
+			rados->objects[i] = NULL;
+			log_err("error creating object.\n");
+			goto failed_obj_create;
+		}
+	}
+
+  return 0;
+
+failed_obj_create:
+	_fio_rados_rm_objects(rados);
+	rados_ioctx_destroy(rados->io_ctx);
+	rados->io_ctx = NULL;
+failed_shutdown:
+	rados_shutdown(rados->cluster);
+	rados->cluster = NULL;
+failed_early:
+	return 1;
+}
+
+static void _fio_rados_disconnect(struct rados_data *rados)
+{
+	if (!rados)
+		return;
+
+	_fio_rados_rm_objects(rados);
+
+	if (rados->io_ctx) {
+		rados_ioctx_destroy(rados->io_ctx);
+		rados->io_ctx = NULL;
+	}
+
+	if (rados->cluster) {
+		rados_shutdown(rados->cluster);
+		rados->cluster = NULL;
+	}
+}
+
+static void fio_rados_cleanup(struct thread_data *td)
+{
+	struct rados_data *rados = td->io_ops_data;
+
+	if (rados) {
+		_fio_rados_disconnect(rados);
+		free(rados->objects);
+		free(rados->aio_events);
+		free(rados);
+	}
+}
+
+static void on_write_op_complete_cb(rados_completion_t cb, void* arg)
+{
+	struct io_u *u = arg;
+	struct fio_rados_iou *fri = u->engine_data;
+	rados_release_write_op(fri->write_op);
+	fri->write_op = NULL;
+
+	rados_aio_release(fri->completion);
+	fri->completion = NULL;
+
+	// mark the pointer to indicate completion for fio_rados_getevents()
+	fri->io_complete = 1;
+
+}
+
+static void on_complete_cb(rados_completion_t cb, void* arg)
+{
+	struct io_u *u = arg;
+	struct fio_rados_iou *fri = u->engine_data;
+
+	rados_aio_release(fri->completion);
+	fri->completion = NULL;
+
+	// mark the pointer to indicate completion for fio_rados_getevents()
+	fri->io_complete = 1;
+}
+
+static int fio_rados_queue(struct thread_data *td, struct io_u *io_u)
+{
+	struct rados_data *rados = td->io_ops_data;
+	struct fio_rados_iou *fri = io_u->engine_data;
+	char* object = rados->objects[io_u->file->engine_pos];
+	int r = -1;
+
+	fio_ro_check(td, io_u);
+
+	fri->io_seen = 0;
+	fri->io_complete = 0;
+
+	if (io_u->ddir == DDIR_WRITE) {
+		r = rados_aio_create_completion(io_u,
+		on_complete_cb,
+		NULL /*cb_safe*/,
+		&fri->completion);
+		if (r < 0) {
+			log_err("rados_aio_create_completion failed.\n");
+			goto failed;
+		}
+
+		r = rados_aio_write(
```

Comment:
Indentation. The fio standard is to use up 80 chars, then spill into the next line if you have more arguments.